core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java

```diff
@@ -446,7 +449,10 @@ boolean isResponseTimedOut(TaskDef taskDefinition, Task task) {
         String taskId = IDGenerator.generate();
         TaskMapperContext taskMapperContext = new TaskMapperContext(workflowDefinition, workflowInstance, taskToSchedule,
                 input, retryCount, retriedTaskId, taskId, this);
-        return taskMappers.get(taskType.name()).getMappedTasks(taskMapperContext);
+        List<Task> tasks = taskMappers.get(taskType.name()).getMappedTasks(taskMapperContext);
+
+        return tasks.stream().filter(task -> executionDAO.getPendingTasksByWorkflow(task.getTaskType(), workflowInstance.getWorkflowId()).stream().
```

Comment:
getPendingTasksByWorkflow is also a very expensive call, it loads all IN_PROGRESS tasks of all workflow executions into the memory then filter by workflowId, please refactor it to get tasks for workflowId first then filter by IN_PROGRESS tasks.

Reply:
Good point, will refactor.